hr_timesheet_sheet/models/account_analytic_line.py

```diff
@@ -50,9 +50,32 @@ def unlink(self):
 
     @api.multi
     def _check_state_on_write(self, values):
-        """ Hook for extensions """
+        """ This method must allow validation of invoices if
+        the timesheet sheet is not in draft state. In sale_timesheet addons,
+        the `_compute_timesheet_revenue` method of `account.invoice` model,
+        modifies the `timesheet_revenue` and `timesheet_invoice_id` fields in
+        analytic lines.
+        """
+        do_check = False
+        for value in values:
+            allowed = self._get_allowed_fields_on_check_state()
+            if value not in allowed:
+                do_check = True
+        if not do_check:
+            return
         self._check_state()
 
+    def _get_allowed_fields_on_check_state(self):
+        """ Extend this method to add fields that could be modified
+        when the timesheet sheet is not in draft.
+        """
+        return [
+            'timesheet_invoice_id',
```

Comment:
I don't thinks this approach (whitelist) is good, especially by whitelisting fields from other modules

Reply:
IMHO, we should have blacklist for `unit_amount`, `project_id`, `employee_id`, `user_id`, and similar instead of whitelist

Reply:
The four fields proposed in that whitelist were required in a production installation.
I don't have any problem to remove two of them, but I guess one day someone will open a new issue to include them again.

Reply:
Ok, were they needed because of sale_timesheet or because any other module/s? This fix should solve just the "sale_timesheet vs hr_timesheet_sheet" problem, leaving the module easily inheritable for any other modules that need it. I haven't digged into it, but I do not understand why would the amount or the uom be changed. Can you pinpoit where are those fields changed?

Reply:
This issue should be addressed in `sale_timesheet_sheet` instead of having a shadow-support-glue within unrelated modules.

Reply:
For me the approach should be: you put the fields that, if modified, are blocked if the sheet is approved. This can be called `blacklist` I think.

If you want to add more fields because a related field shouldn't be changed (which is very weird), then:

1. The module adding the field depends on this one, and overwrite the method that returns the blacklist adding this new field.
2. As the blacklist is a list of strings, you can add the field in the main module if inheritance is not possible, adding a comment explaining the situation. While the modules is an OCA one, this is allowed.

Reply:
The standard `hr_timesheet_sheet` in Odoo 10.0 [blocks every field](https://github.com/odoo/odoo/blob/10.0/addons/hr_timesheet_sheet/models/account_analytic_line.py#L49). Here we are making an exception for the invoice validation, but the approval of a timesheet sheet is meant to freeze all the values of its lines. It was always working that way.

Can you imagine how many discussions between employees and employers would occur in a company, in case there's just a suspect that the timesheet lines can be modified after the sheet validation?

No way. I keep my solution. Feel free to propose an alternative solution in a new PR.

Reply:
Well, some modules Odoo aren't the best example of proactive providing hooks for extensibility. IMHO, this module should list all _known_ fields to be protected, that will provide same result as right now. At the same time, it won't interfere with any other modules that work with AALs

Reply:
Please go ahead.

Reply:
@astirpe if you look to the enterprise version now (`timesheet_grid`), this restriction is no more there as very... restrictive. I think the whitelist approach will generate a lot of undesired side effects as ORM constantly can update the record without being a "malicious" attempt of changing some validated data. At the end, what you are validating in the sheet is the fields you are seeing there: user, time amount, date, project, task and little more (maybe few more if new extension modules are put, but then they can add it to the blacklist), but nothing with message_ids, computed fields, and so on, that are the conflicting ones that are updated through other automations. Can you please reconsider it?

Reply:
Something like #194